MatrixSDK/Background/MXBackgroundSyncService.swift

```diff
@@ -0,0 +1,458 @@
+// 
+// Copyright 2020 The Matrix.org Foundation C.I.C
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+// http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+//
+
+import Foundation
+
+/// Errors can be raised by `MXBackgroundSyncService`.
+public enum MXBackgroundSyncServiceError: Error {
+    case unknown
+    case unknownAlgorithm
+    case decryptionFailure
+}
+
+/// This class can be used to sync in background, keeping the user offline. It does not initiate MXSession or MXCrypto instances.
+/// Sync results are written to a SyncResponseFileStore.
+@objcMembers public class MXBackgroundSyncService: NSObject {
+    
+    private enum Queues {
+        static let processingQueue: DispatchQueue = DispatchQueue(label: "MXBackgroundSyncServiceQueue")
+        static let dispatchQueue: DispatchQueue = .main
+    }
+    
+    private enum Constants {
+        static let syncRequestServerTimout: UInt = 0
+        static let syncRequestClientTimout: UInt = 20 * 1000
+        static let syncRequestPresence: String = "offline"
+    }
+    
+    private let credentials: MXCredentials
+    private let syncResponseStore: SyncResponseStore
+    private let store: MXStore
+    private let cryptoStore: MXCryptoStore
+    private let olmDevice: MXOlmDevice
+    private let restClient: MXRestClient
+    private var pushRulesManager: MXBackgroundPushRulesManager
+    
+    /// Cached events. Keys are even identifiers.
+    private var cachedEvents: [String: MXEvent] = [:]
+    
+    /// Initializer
+    /// - Parameter credentials: account credentials
+    public init(withCredentials credentials: MXCredentials) {
+        self.credentials = credentials
+        syncResponseStore = SyncResponseFileStore()
+        syncResponseStore.open(withCredentials: credentials)
+        restClient = MXRestClient(credentials: credentials, unrecognizedCertificateHandler: nil)
```

Comment:
You may want
`restClient.completionQueue = Queues. processingQueue'
to have response on the right queue.
By default, they return on the main queue.